Cython/Compiler/Nodes.py

```diff
@@ -140,6 +140,24 @@ def call(*args, **kwargs):
         return super().__new__(cls, name, bases, attrs)
 
 
+class CopyWithUpTreeRefsMixin:
+    def __deepcopy__(self, memo):
+        # Any references to objects further up the tree should not be deep-copied.
+        # However, if they're in memo (because they've already been deep-copied because
+        # we're copying from far enough up the tree) then they should be replaced
+        # with the memorised value.
+
+        cls = self.__class__
+        result = cls.__new__(cls)
+        for k, v in self.__dict__.items():
+            if k in self.uptree_ref_attrs:
+                # Note that memo being keyed by "id" is a bit of an implementation detail;
+                # the documentation says to treat it as opaque.
+                v = memo.get(id(v), v)
+            setattr(result, k, v)
+        return result
```

Comment:
Hmm, ok – I know, I linked to the answer that seemed to match my intentions, but didn't actually see the memo part in it. Sure, you're right, if there is a replacement already, then it must be used. And the documentation is clear about about `memo` being opaque. :-/

I guess, either this or your previous code will solve the problem, both being a bit fragile in the sense that they tightly couple with the `deepcopy()` implementation. I don't see a way to avoid that, so I'll leave the choice between the two approaches to you. Sorry for being a bit short-sighted here.

Reply:
I've at least checked CPython, PyPy and GraalPython and they all use the same implementation of deepcopy so it should at least work reasonably widely.

I think it'd be possible to come up with an alternative if we have to if it does change in future. But it isn't particularly obvious...

I'll use the current version for now I think.